File: php/src/Runtime.php
<?php

declare(strict_types=1);

namespace LiteDbModel\Runtime;

/**
 * litedbmodel v2 SCP — PHP runtime (WS7d, #33; static-makeSQL flip, epic #43/#45).
 *
 * The PHP leg of the litedbmodel v2 SCP multi-language runtime. It interprets the
 * language-neutral §8 published STATIC makeSQL `SqlBundle` (a read `readGraph` = `compileBehaviors`'
 * REAL `Select`/map `ComponentGraphIR` + per-node STATIC statement templates, or a write
 * `transaction` plan of gate-first makeSQL statements, dialect-tagged) and executes it against
 * a PDO SQL driver, semantics-identical to the TS reference (`src/scp/runtime.ts` +
 * `src/scp/makesql/*`).
 *
 * It is a THIN runtime: the read-graph execution (plan-stage / map iteration / wiring / Φ output
 * assembly) is the NATIVE walker in {@link StaticBundle} (#12 — no bc `runBehavior`); only the
 * deferred Expression-IR value-specs + skip are evaluated by the VENDORED behavior-contracts PHP
 * port (`src/BehaviorContracts/` `ExprEval`); the static makeSQL assemble/render/execute lives in
 * {@link StaticBundle} + {@link WriteRuntime}. This runtime is the thin FACADE that dispatches a
 * bundle to the read graph executor or the gate-first write transaction. The reduced
 * fragment-tree render path (`Render::renderOperation`) is RETIRED for the SQL path — makeSQL is
 * the sole read/render path.
 *
 * ## Value model (byte-identity with the reference)
 *
 * The bundle JSON is decoded with `json_decode($json, false)` so IR nodes are `stdClass` and
 * lists are PHP arrays — exactly what the vendored bc `ExprEval`/`Behavior` expect. SQL rows
 * are returned to bc as `stdClass` so that (a) a map element's `{ref:[$e, col]}` resolves via
 * property access and (b) the Φ `{obj:{…}}` output JSON-encodes to the same object shape as the
 * TS reference. bc evaluates integer params to PHP `int` (PDO binds them directly).
 *
 * ## Dialect / driver seam
 *
 * The conformance bar executes against a REAL in-process PDO SQLite database. The dialect axis
 * (`sqlite`/`postgres`/`mysql`) governs the rendered SQL TEXT (`?`→`$N` for Postgres); the
 * seeded DB is SQLite regardless — so a PG/MySQL-tagged bundle's rendered params still bind
 * positionally, and the EXECUTED result is dialect-invariant (the §10 promise). Live PG/MySQL
 * PDO execution is DEFERRED to a coordinated cross-language docker pass; the handler seam takes
 * any PDO connection, so pgsql/mysql plug in unchanged when that pass runs.
 */
final class Runtime
{
    /** Version mirrored from package.json by scripts/sync-versions.mjs (SSoT). */
    public const VERSION = '2.1.0';

    /**
     * Render the PRIMARY read node's statements of a `ReadGraph` against a scope for a dialect →
     * ['sql'=>…, 'params'=>…] (static-bundle.ts `renderReadPrimary`). The render axis for the
     * conformance golden.
     *
     * @param \stdClass $readGraph the compiled ReadGraph (json_decode(.., false) shape).
     * @param array<string,mixed> $scope
     * @return array{sql:string, params:list<mixed>}
     */
    public static function renderReadPrimary(\stdClass $readGraph, array $scope): array
    {
        return StaticBundle::renderReadPrimary($readGraph, $scope);
    }

    /** The dialect NULLS-ordering primitive (dialect.ts `orderByNulls`). */
    public static function orderByNulls(string $expr, string $dir, string $nulls, string $dialect): string
    {
        return Dialect::forName($dialect)->orderByNulls($expr, $dir, $nulls);
    }

    /**
     * Execute a §8 read/exec `SqlBundle` end-to-end (runtime.ts `executeBundle`): delegate to
     * {@link StaticBundle::executeReadGraph}, the NATIVE read-graph walker (#12 — NO bc
     * `runBehavior`) which owns the plan / map / wire / output orchestration itself, renders each
     * node's static statement templates against the walk scope, and runs REAL SQL via PDO.
     * Consumes ONLY the serialized bundle + vendored bc `ExprEval` (deferred params + skip) —
     * never re-deriving litedbmodel's Backend-Compile.
     *
     * @param \stdClass $bundle the §8 published bundle (json_decode(.., false) shape).
     * @param array<string,mixed> $input the bound input scope.
     * @return mixed the component's Φ output.
     */
    public static function executeBundle(\stdClass $bundle, array $input, \PDO $db): mixed
    {
        $readGraph = $bundle->readGraph ?? null;
        if (!($readGraph instanceof \stdClass)) {
            throw new \RuntimeException(
                "scp runtime: bundle '" . (string) ($bundle->name ?? '')
                . "' carries no read graph (single-statement writes ride the write path)"
            );
        }
        try {
            return StaticBundle::executeReadGraph($readGraph, $input, $db);
        } catch (\Throwable $e) {
            throw self::reErrorToSqlFailure($e);
        }
    }

    /**
     * Execute a §8 write-tx `SqlBundle`'s derived transaction plan as ONE real transaction with
     * gate-first short-circuit (runtime.ts `executeTransactionBundle` + write-runtime.ts
     * `executeTransaction`). Consumes ONLY the serialized `TransactionPlan` (pure JSON) + the
     * render pipeline + PDO.
     *
     * @param \stdClass $bundle
     * @param array<string,mixed> $input
     * @return array{committed:bool, shortCircuit?:array{statementId:string,reason:string}, entity:?array<string,mixed>, executed:list<string>}
     */
    public static function executeTransactionBundle(\stdClass $bundle, array $input, \PDO $db): array
    {
        $plan = $bundle->transaction ?? null;
        if (!($plan instanceof \stdClass)) {
            throw new \RuntimeException(
                'scp write: this bundle carries no transaction plan (not a write-time-relations Command bundle)'
            );
        }
        return WriteRuntime::executeTransaction($db, $plan, $input, Dialect::forName((string) $bundle->dialect));
    }

    /**
     * If a `runPlan` `OP_FAILED` carries a mapped-failure message, re-surface the structured
     * {@link SqlFailure} (the message embeds the original SQLite code). Non-driver errors are
     * re-thrown verbatim (runtime.ts `reErrorToSqlFailure`).
     */
    private static function reErrorToSqlFailure(\Throwable $e): \Throwable
    {
        $message = $e->getMessage();
        if (preg_match('/(SQLITE_[A-Z_]+)/', $message, $m) === 1) {
            return SqlFailure::fromCode($m[1], $message);
        }
        return $e;
    }
}

File: go/litedbmodel_runtime/runtime.go
// Package litedbmodel_runtime is the Go leg of the litedbmodel v2 SCP multi-language runtime
// (WS7c, #32).
//
// It consumes the language-neutral §8 published bundle (SqlBundle: sql + fragment tree +
// closed-set Expression-IR param slots + transaction plan, dialect-tagged) and executes it against
// a database/sql driver, semantics-identical to the TS reference (src/scp). It is NATIVE: the
// read-graph orchestration (which node runs, map iteration, wire binding, output assembly) is a
// CLOSED-SET native walker (executeReadGraphNative) — NOT bc.RunBehavior (the generic IR
// interpreter, retired). The only bc dependency on the exec path is the per-statement typed-param
// evaluation (bc.EvaluateExpression resolves each deferred `{ref:…}`/`coalesce`/`__jsonArray` slot);
// every statement's SQL is fixed text carried verbatim.
//
// Execution pipeline (spec §3):
//
//	validate → fragment select (SKIP) → array expand → param eval + bind → SQL execute → native assembly
//
// #12: the read graph carries compileBehaviors' REAL Select/Count/map nodes; the native walker
// renders each node's pre-compiled statementsById against the walk scope directly (no `__scope`).

package litedbmodel_runtime

import (
	"fmt"
	"regexp"
	"strings"

	bc "github.com/foo-ogawa/behavior-contracts/go"
)

// Version is synced from package.json by scripts/sync-versions.mjs (Go = VCS tag, not a manifest
// field, so this constant is the in-source mirror the CI tag check compares against).
const Version = "2.0.2"

// entityRoot is the body-write RETURNING row exposed to later tx stages under `$.entity.*`
// (writes.ts ENTITY_ROOT).
const entityRoot = "__entity"

var returningRe = regexp.MustCompile(`(?i)\breturning\b`)

// SqlBundle is the parsed makeSQL published bundle. A READ bundle carries a `readGraph` (the
// REAL Select-node ComponentGraphIR + per-node static statement templates); a WRITE bundle carries a
// gate-first `transaction` plan. The heavy fields stay as raw bc JNodes so their Expression-IR slots
// are evaluated by bc (not pre-decoded), byte-true to the TS SqlBundle.
type SqlBundle struct {
	Dialect     string
	Name        string
	ReadGraph   *ReadGraph // read/exec bundle: the static makeSQL read graph (nil for write bundles)
	Transaction *bc.JObj   // write-tx plan (nil for read/exec bundles)
}

// ParseBundle parses a makeSQL bundle from its pure-JSON bytes (the published artifact) into a
// SqlBundle whose IR/param slots stay as bc JNodes. This is the "executes from published JSON
// alone" entry point: no TS state, no re-derivation.
func ParseBundle(data []byte) (*SqlBundle, error) {
	root, err := bc.ParseJSONOrdered(data)
	if err != nil {
		return nil, fmt.Errorf("scp runtime: bundle parse: %w", err)
	}
	obj, ok := root.(*bc.JObj)
	if !ok {
		return nil, fmt.Errorf("scp runtime: bundle root is not an object")
	}
	return BundleFromJObj(obj)
}

// BundleFromJObj builds a SqlBundle from a parsed bundle object (shared by ParseBundle + the
// vector runner, which parses whole suites at once and hands the per-vector `bundle` object here).
func BundleFromJObj(obj *bc.JObj) (*SqlBundle, error) {
	b := &SqlBundle{}
	if d, ok := obj.Get("dialect"); ok {
		b.Dialect, _ = d.(string)
	}
	if n, ok := obj.Get("name"); ok {
		b.Name, _ = n.(string)
	}
	if rgN, ok := obj.Get("readGraph"); ok {
		if rg, ok := rgN.(*bc.JObj); ok {
			g, err := ReadGraphFromJObj(rg)
			if err != nil {
				return nil, err
			}
			b.ReadGraph = g
		}
	}
	if tx, ok := obj.Get("transaction"); ok {
		b.Transaction, _ = tx.(*bc.JObj)
	}
	return b, nil
}

// ── Public runtime entrypoint ─────────────────────────────────────────────────

// ExecuteBundle executes a read/exec SqlBundle end-to-end (runtime.ts executeBundle): a read bundle
// carries a `readGraph` (the REAL Select-node ComponentGraphIR + static statements); a CLOSED-SET
// native walker drives map/Φ/wiring (never bc.RunBehavior) and renders + executes each node against
// REAL SQL. This is the SAME code path a consumer runtime follows — it consumes ONLY the serialized
// bundle + bc runtime-core, never re-running litedbmodel's Backend-Compile.
func ExecuteBundle(bundle *SqlBundle, input *bc.Obj, db SQLDB) (bc.Value, error) {
	if bundle.ReadGraph == nil {
		return nil, fmt.Errorf("scp runtime: bundle '%s' carries no read graph (single-statement writes ride the write path)", bundle.Name)
	}
	return ExecuteReadGraph(bundle.ReadGraph, input, db)
}

// reErrorToSqlFailure re-surfaces a structured SqlFailure from a bc OP_FAILED whose message embeds
// a `SQLITE_*` code (runtime.ts reErrorToSqlFailure). Non-driver errors are returned verbatim.
func reErrorToSqlFailure(err error) error {
	msg := err.Error()
	if i := strings.Index(msg, "SQLITE_"); i >= 0 {
		rest := msg[i:]
		end := 0
		for end < len(rest) && (rest[end] == '_' || (rest[end] >= 'A' && rest[end] <= 'Z')) {
			end++
		}
		code := rest[:end]
		switch {
		case code == "SQLITE_CONSTRAINT_FOREIGNKEY":
			return &SqlFailure{Kind: KindForeignKeyViolation, Policy: "fail", SqliteCode: code, Msg: msg}
		case strings.HasPrefix(code, "SQLITE_CONSTRAINT"):
			return &SqlFailure{Kind: KindConstraintViolation, Policy: "fail", SqliteCode: code, Msg: msg}
		case code == "SQLITE_BUSY" || code == "SQLITE_LOCKED":
			return &SqlFailure{Kind: KindRetryable, Policy: "retry", SqliteCode: code, Msg: msg}
		default:
			return &SqlFailure{Kind: KindDriverError, Policy: "fail", SqliteCode: code, Msg: msg}
		}
	}
	return err
}
